lib/iris/_shapefiles.py

```diff
@@ -3,241 +3,447 @@
+    geometry: shapely.Geometry,
+    minimum_weight: float,
+) -> np.array:
+    """Get a mask based on the geometry and minimum weight.
 
     Parameters
     ----------
     cube : :class:`iris.cube.Cube`
+        The cube to mask.
+    geometry : :class:`shapely.Geometry`
+        The geometry to use for masking.
+    minimum_weight : float
+        The minimum weight of the geometry to be included in the mask.
+        If the weight is less than this value, the geometry will not be
+        included in the mask.
 
     Returns
     -------
-    tuple of str
-        The names of the primary latitude and longitude coordinates.
-
+    :class:`np.array`
+        An array of the shape of the x & y coordinates of the cube, with points
+        to mask equal to True.
     """
-    latc = (
-        cube.coords(axis="y", dim_coords=True)[0]
-        if cube.coords(axis="y", dim_coords=True)
-        else -1
-    )
-    lonc = (
-        cube.coords(axis="x", dim_coords=True)[0]
-        if cube.coords(axis="x", dim_coords=True)
-        else -1
-    )
-
-    if -1 in (latc, lonc):
-        msg = "Error retrieving 1d xy coordinates in cube: {!r}"
-        raise ValueError(msg.format(cube))
+    if not cube.coord(axis="X", dim_coords=True).has_bounds():
+        cube.coord(axis="X", dim_coords=True).guess_bounds()
+    if not cube.coord(axis="Y", dim_coords=True).has_bounds():
+        cube.coord(axis="Y", dim_coords=True).guess_bounds()
+    # Get the shape of the cube
+    w = len(cube.coord(axis="X", dim_coords=True).points)
+    h = len(cube.coord(axis="Y", dim_coords=True).points)
+    # Get the bounds of the cube
+    # x_bounds = _get_mod_rebased_coord_bounds(cube.coord(x_name))
+    # y_bounds = _get_mod_rebased_coord_bounds(cube.coord(y_name))
+    x_bounds = cube.coord(axis="X", dim_coords=True).bounds
+    y_bounds = cube.coord(axis="Y", dim_coords=True).bounds
+    # Generate Sort-Tile-Recursive (STR) packed R-tree of bounding boxes
+    # https://shapely.readthedocs.io/en/stable/strtree.html
+    grid_boxes = [
+        sgeom.box(x[0], y[0], x[1], y[1]) for y, x in product(y_bounds, x_bounds)
+    ]
+    grid_tree = shapely.STRtree(grid_boxes)
+    # Find grid boxes indexes that intersect with the geometry
+    idxs = grid_tree.query(geometry, predicate="intersects")
+    # Get grid box indexes that intersect with the minimum weight criteria
+    mask_idxs_bool = [
+        grid_boxes[idx].intersection(geometry).area / grid_boxes[idx].area
+        >= minimum_weight
+        for idx in idxs
+    ]
+    mask_idxs = idxs[mask_idxs_bool]
+    mask_xy = [list(product(range(h), range(w)))[i] for i in mask_idxs]
+    # Create mask from grid box indexes
 
-    latitude = latc.name()
-    longitude = lonc.name()
-    return latitude, longitude
+    weighted_mask_template = np.ones((h, w), dtype=bool)
+    # Set mask = True for grid box indexes identified above
+    for xy in mask_xy:
+        weighted_mask_template[xy] = False
```

Comment:
If you're concerned about performance, I believe there are ways to refactor this to assign `False` in a single NumPy operation instead of a loop.

If you're not worried then I'm not worried.

Reply:
Not worried, but have tweaked slightly.